### PR TITLE
Plane: allow fallback to DCM in VTOL modes

### PR DESCRIPTION
**Pull request by tridge**
this adds the necessary functions to allow for flying a quadplane on DCM as an emergency fallback. It sets the NavGainScalar to 0.5 to reduce the VTOL controller gains to allow planes to cope with the higher lag of DCM
Tested in RealFlight only so far
Thanks to @lthall for assistance with this PR

Test log on a Foxtech HWing here:
http://uav.tridgell.net/HWing/2021-08-15/log48.bin

Video of the test here:
https://www.youtube.com/watch?v=ZI65d29dIXs

### CODE REVIEW

File: libraries/AC_AttitudeControl/AC_PosControl.cpp [hunk 1]
```diff
@@ -962,6 +962,7 @@ void AC_PosControl::update_z_controller()
     float pos_target_zf = _pos_target.z;
 
     _vel_target.z = _p_pos_z.update_all(pos_target_zf, curr_alt, _limit.pos_down, _limit.pos_up);
+    _vel_target.z *= AP::ahrs().getEkfControlScaleZ();
```
Comment (lthall) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp:965]: I would like to see the variable ekfNavVelGainScaler changed to be consistent with this variable ie getEkfControlScaleXY.

**Review by lthall — CHANGES_REQUESTED** [review]

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp:965]: done

**Review by lthall — APPROVED** [review]

File: libraries/AC_AttitudeControl/AC_PosControl.cpp [hunk 2]
```diff
@@ -625,16 +625,16 @@ void AC_PosControl::update_xy_controller()
     }
     _last_update_xy_us = AP_HAL::micros64();
 
-    float ekfGndSpdLimit, ekfNavVelGainScaler;
-    AP::ahrs().getEkfControlLimits(ekfGndSpdLimit, ekfNavVelGainScaler);
+    float ekfGndSpdLimit, ekfControlScaleXY;
+    AP::ahrs().getEkfControlLimits(ekfGndSpdLimit, ekfControlScaleXY);
```
Comment (peterbarker) [inline libraries/AC_AttitudeControl/AC_PosControl.cpp:629]:
```suggestion
    float ahrsGndSpdLimit, ahrsControlScaleXY;
    AP::ahrs().getEkfControlLimits(ahrsGndSpdLimit, ahrsControlScaleXY);
```

If we're going to "fix" the naming then let's fix it.

File: libraries/AC_WPNav/AC_Loiter.cpp [hunk 3]
```diff
@@ -208,8 +208,8 @@ void AC_Loiter::sanity_check_params()
 ///		updated velocity sent directly to position controller
 void AC_Loiter::calc_desired_velocity(float nav_dt, bool avoidance_on)
 {
-    float ekfGndSpdLimit, ekfNavVelGainScaler;
-    AP::ahrs().getEkfControlLimits(ekfGndSpdLimit, ekfNavVelGainScaler);
+    float ekfGndSpdLimit, ekfControlScaleXY;
+    AP::ahrs().getEkfControlLimits(ekfGndSpdLimit, ekfControlScaleXY);
```
Comment (peterbarker) [inline libraries/AC_WPNav/AC_Loiter.cpp:212]:
```suggestion
    AP::ahrs().getControlLimits(ekfGndSpdLimit, ekfControlScaleXY);
```

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 4]
```diff
@@ -2124,12 +2131,27 @@ void AP_AHRS::getEkfControlLimits(float &ekfGndSpdLimit, float &ekfNavVelGainSca
...
+        // no limit on gains, large vel limit
+        ekfGndSpdLimit = 400;
+        ekfNavVelGainScaler = 1;
         break;
 #endif
     }
 }
 
+/*
+  get gain recuction factor for Z controllers
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS.cpp:2143]:
```suggestion
  get gain factor for Z controllers
```

File: libraries/AP_AHRS/AP_AHRS.cpp [hunk 5]
```diff
@@ -2627,10 +2649,9 @@ void AP_AHRS::set_terrain_hgt_stable(bool stable)
 //  returns true on success (i.e. the EKF knows it's latest position), false on failure
 bool AP_AHRS::get_location(struct Location &loc) const
 {
-    switch (ekf_type()) {
+    switch (active_EKF_type()) {
     case EKFType::NONE:
-        // We are not using an EKF so no data
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS.cpp:2632]: method description needs changing now

File: libraries/AP_AHRS/AP_AHRS.h [hunk 6]
```diff
@@ -188,7 +188,8 @@ class AP_AHRS : public AP_AHRS_DCM {
     void writeExtNavVelData(const Vector3f &vel, float err, uint32_t timeStamp_ms, uint16_t delay_ms) override;
 
     // get speed limit
-    void getEkfControlLimits(float &ekfGndSpdLimit, float &ekfNavVelGainScaler) const;
+    void getEkfControlLimits(float &ekfGndSpdLimit, float &controlScaleXY) const;
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS.h:191]:
```suggestion
    void getControlLimits(float &gndSpdLimit, float &controlScaleXY) const;
```

I'd like to split this call at some stage.  Several callers only want one of these.

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp [hunk 7]
```diff
@@ -1047,10 +1051,13 @@ bool AP_AHRS_DCM::get_position(struct Location &loc) const
     loc.relative_alt = 0;
     loc.terrain_alt = 0;
     loc.offset(_position_offset_north, _position_offset_east);
-    if (AP::ahrs().get_fly_forward() && _have_position) {
-        float gps_delay_sec = 0;
-        gps.get_lag(gps_delay_sec);
-        loc.offset_bearing(gps.ground_course(), gps.ground_speed() * gps_delay_sec);
+    if (_have_position) {
+        uint32_t now = AP_HAL::millis();
```
Comment (peterbarker) [inline libraries/AP_AHRS/AP_AHRS_DCM.cpp:1055]:
```suggestion
        const uint32_t now = AP_HAL::millis();
```

had to be said

**Review by peterbarker — CHANGES_REQUESTED** [review]
I did not repeat comments made on the other PR.

I'm still concerned about origin shifting about.

File: libraries/AC_AttitudeControl/AC_PosControl.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AC_AttitudeControl/AC_PosControl.cpp]: done

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_AHRS/AP_AHRS.cpp]: done

File: libraries/AP_AHRS/AP_AHRS.cpp (hunk 5 above)
Comment (tridge) [reply libraries/AP_AHRS/AP_AHRS.cpp:2632]: done

File: libraries/AP_AHRS/AP_AHRS.h (hunk 6 above)
Comment (tridge) [reply libraries/AP_AHRS/AP_AHRS.h]: done

File: libraries/AP_AHRS/AP_AHRS_DCM.cpp (hunk 7 above)
Comment (tridge) [reply libraries/AP_AHRS/AP_AHRS_DCM.cpp]: done

File: libraries/AC_WPNav/AC_Loiter.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AC_WPNav/AC_Loiter.cpp]: done

**Review by peterbarker — APPROVED** [review]
LGTM.

**Pull request merged**